### PR TITLE
Give default foreground to dark/light plus themes

### LINKED ISSUE #45: dark_plus theme colors braces black
Not much else to say... This also seems to happen with dark_vs. I spent a bit of time digging into it, but haven't figured out the fix.

image

### PR DESCRIPTION
**Pull request by octref**
This PR fixes shikijs/shiki#45

Themes should set fallback foreground color with the first element of `tokenColors`. For example, https://github.com/microsoft/vscode/blob/c5fe5d7c1a1bb69ef83ac9064bea3210642d0cc4/extensions/theme-solarized-dark/themes/solarized-dark-color-theme.json#L4-L7

Also makes `themes.ts` consistent with the JSON themes.

I didn't make changes to `colorRegistry.ts`, but it seems it's inconsistent with these two as well...

/cc @misolori 👋

### CODE REVIEW

File: extensions/theme-defaults/themes/light_defaults.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
 	"$schema": "vscode://schemas/color-theme",
 	"name": "Light Default Colors",
 	"colors": {
-		"editor.background": "#FFFFFF",
+		"editor.background": "#FFFFFE",
```
Comment (misolori) [inline extensions/theme-defaults/themes/light_defaults.json:5]: Did this accidentally get changed?

Comment (octref) [reply extensions/theme-defaults/themes/light_defaults.json:5]: I thought it's canonical:
https://github.com/microsoft/vscode/blob/196500f3647d9e26b71054652ae79e7886803117/src/vs/editor/standalone/common/themes.ts#L15
https://github.com/microsoft/vscode/blob/196500f3647d9e26b71054652ae79e7886803117/src/vs/editor/standalone/common/themes.ts#L71

Comment (misolori) [reply extensions/theme-defaults/themes/light_defaults.json:5]: Ah, thanks for clarifying and for fixing it 😉

**Review by miguelsolorio — APPROVED** [review]
Tried this out in master and didn't see any significant impact. Leaving it to @aeschli for final approval.

**Comment by aeschli** [comment]
The problem is that there are two places to define the editor fore-/background colors:
- `editor.foreground`/`editor.background` in the workbenchColors 
- rule with no scope in TextMate rule settings.

To resolve that, we massage the TextMate rule settings before passing to the TextMate engine:
Put the `editor.foreground`/`editor.background` colors first, filter all rules with no scope.

https://github.com/microsoft/vscode/blob/dd85f62e65325be84c2da2a63943a3322c3fb6a2/src/vs/workbench/services/themes/common/colorThemeData.ts#L115

So I'd rather not take the PR.

**Comment by octref** [comment]
This assumes theme is only for usage in vscode. When using vscode-textmate standalone, it does not handle `editor.foreground` or `editor.background`, but it does handle the first setting without a scope: 

https://github.com/microsoft/vscode-textmate/issues/131

https://github.com/microsoft/vscode-textmate/blob/ebcfe99ce2d262165c53e44db71575f37750c070/src/theme.ts#L185-L196

In the case of vscode-textmate, it's only trying to find default colors for tokenization. Making `colors.editor.foreground` the defualt color for tokenization doesn't make sense to me.

**Comment by alexdima** [comment]
> In that case I'd say the dark-plus theme is not complete, since it doesn't set a default foreground, which should be set to white to avoid falling back to black.

@aeschli I agree with @octref . AFAIK defaults should be set via a rule with an empty scope. If we want our dark-plus theme to be *portable* and make sense also outside vscode, then I think it should have a rule with an empty scope in order to define the defaults.

**Comment by aeschli** [comment]
The token scopes that are passed to vscode-textmate are complete and standard tmTheme data, including the default fore and background colors using the empty scope. Check the code pointer that generates that tmTheme data from our theme.
That processing step is needed as the editor fore/background color can also come from user colors in the settings.

To me it just adds to confusion if users would add a tokenScope rule for the foreground that then later is just filtered. So I better not have it in there

If you want to use our themes you need to make the same processing steps and understand our theme format.

**Comment by octref** [comment]
https://github.com/microsoft/vscode/blob/94b05d07cb2891f7d42f51c217ae2ffb92e892eb/extensions/theme-solarized-dark/themes/solarized-dark-color-theme.json#L364
https://github.com/microsoft/vscode/blob/94b05d07cb2891f7d42f51c217ae2ffb92e892eb/extensions/theme-solarized-light/themes/solarized-light-color-theme.json#L362

**Pull request closed without merging**